pkg/wrangler/capi.go

```diff
@@ -164,18 +221,42 @@ func (d *DeferredCAPIRegistration) DeferFuncWithError(clients *Context, f func(w
 	return errChan
 }
 
+// DeferRegistration enqueues a function to be executed once the CAPI CRDs are available by adding it to the registration function pool.
+// The functions passed to DeferRegistration are expected to register one or more event handlers which rely on CAPI clients.
+// Calls to DeferRegistration are processed in the order they are made. Calls to DeferRegistration made after the CAPI CRDs are
+// available will execute immediately, and the controller factory will be immediately started.
 func (d *DeferredCAPIRegistration) DeferRegistration(ctx context.Context, clients *Context, register func(ctx context.Context, clients *Context) error) error {
-	d.lock.Lock()
-	defer d.lock.Unlock()
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
 	d.wg.Add(1)
-	if d.CAPIInitialized {
-		return clients.StartSharedFactoryWithTransaction(ctx, func(ctx context.Context) error {
+
+	if d.CAPIInitComplete {
+		logrus.Debugf("[deferred-capi - DeferRegistration] Executing deferred registration function as CAPI is initilized")
+		defer func() {
+			logrus.Debugf("[deferred-capi - DeferRegistration] deferred registration function has completed")
+		}()
+
+		clients.controllerLock.Lock()
```

Comment:
maybe place this block into an unnamed nested func to be able to use lock/defer unlock

Reply:
Good idea